trade-finance-manager-api/api-tests/types/test-user.ts

```diff
@@ -0,0 +1,16 @@
+import { TeamId } from '@ukef/dtfs2-common';
+
+/**
+ * Users returned by the test user fluent builder
+ */
+export type TestUser = {
+  _id?: string;
+  username: string;
+  password: string;
+  email: string;
+  teams: TeamId[];
+  timezone: string;
+  firstName: string;
+  lastName: string;
+  token: string;
+};
```

Comment:
Do we already have a `User` type that could be used here instead of a new `TestUser` type?

If there's an existing `User` type, but the tests require additional fields, perhaps we could extend the test type instead

Reply:
Not sure what you mean by hacky - do we have an existing `User` type? I can't seem to find 

Just want to avoid having duplicate or almost identical types when they could have a common interface

Reply:
Can you see the latest version of the file? I think there is a case for separate types for those stored in the database (database models) and when they've been serialised (Data Transfer Objects - i think).

Happy to somewhat commonise them here if you think thats best

Reply:
I've hit this type issue on SSO (I'll bring it up tomorrow in standup as we need to decide an agreed pattern).

We should have the following types for any given thing ideally:
- Database model
- Internal service use type (can be the same as DTO)
- Data transfer object (as it leaves our service)

Then we'll need to have an incoming DTO for the service recieving the object (basically a serialised DTO object), a mapper that maps the serialised DTO either back to the DTO object or return a new, leaner object (the leaner object is slightly pointless imo).

As we pass through multiple services, we might need more than one DTO for an object (for instance, UI layer should only recieve things that the UI cares about, but the API layer might care about more things depending on responsibility of services, a good example of this would be `facilities` or `deals` etc -- we wouldn't want to return things to the UI layer, but we would want to be able to pass them around our 
API services, so the need for multiple DTOs could be present).

Commonisation is interesting, as the commonisation appears to have occured prior to libs common, and may already be out of sync (they are on SSO ticket).

I've seen a libs common and a TFM API `TfmUser` model for instance.

Regardless, my understanding so far is that we have a `TfmUser` db model -- and we should, on all repo calls, be transforming that type into another type, removing information like salt and hash etc (ideally we would do this in the mongo queries themselves). At the minute, my understanding is this is pretty much `TfmSessionUser` object for all intents and purposes.

Reply:
Good reply @AlexBramhill 

With my current knowledge/understanding, I don't see why a model should be different throughout various areas. IMO they should ideally be aligned, avoiding mapping, multiple types, potential confusion etc.

Appreciate there are scenarios where they _should_ be different - e.g for a user, probably only the API/DB should care about e.g salt/hash fields and the UI shouldn't consume or know about such fields. So in this scenario there could be a common user type and the API/DB extends the type.

This feels like a separate thing to tackle outside of this PR and sounds like it doesn't affect just users.

Shall we setup a call to talk about this and agree on a pattern like @AlexBramhill said?

Reply:
Would be great to have a call about this sometime :) 

I definitely agree that we should be avoiding mapping these objects around too much, but unfortunately theres a lot of that already.
The issue is particularly visible when you look at the objects being consumed by the ui (e.g. I created a front end facility type `gef-ui\server\types\facility.ts` because the final object is so different to `libs\common\src\types\mongo-db-models\facility.ts`. The UnixTimestampStrings are mapped to IsoDateTimeStamp somewhere, ObjectIds are serialised to strings)

Reply:
If we did have the ideal world where things weren't mapped, we could use a [Jsonify](https://effectivetypescript.com/2020/04/09/jsonify/) type helper to get around the serialisation issue

Reply:
Resolving but will chat about the plan for this going forward at some point :)